Add global setup to timer to allow #include, etc.

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53296 Definition infrastructure for instruction count ubenchmarks
* #53295 Add repeats to Timer.collect_callgrind(...)
* #53294 Quality of life improvements to Timer
* **#53293 Add global setup to timer to allow #include, etc.**

Instruction count benchmarks need some includes for IValues, but this is also just generally useful. (Unlike Python where you can just drop imports anywhere, C++ will get very upset if you `#include` in a function body...)

Differential Revision: [D26906684](https://our.internmc.facebook.com/intern/diff/D26906684)

Commit:
Add global setup to timer to allow #include, etc.

[ghstack-poisoned]

Commit:
Update on "Add global setup to timer to allow #include, etc."




[ghstack-poisoned]

Commit:
Update on "Add global setup to timer to allow #include, etc."




[ghstack-poisoned]

robieta commented:
Comments from https://github.com/pytorch/pytorch/pull/50604:
https://github.com/pytorch/pytorch/pull/50604#discussion_r566169317 I've started pushing more and more things to be kwarg only. This was a good tip.

https://github.com/pytorch/pytorch/pull/50604#discussion_r566169749 It would, but you don't need it. (The Pythonic thing is to put it in `setup`.)

Commit:
Update on "Add global setup to timer to allow #include, etc."


Instruction count benchmarks need some includes for IValues, but this is also just generally useful. (Unlike Python where you can just drop imports anywhere, C++ will get very upset if you `#include` in a function body...)

[ghstack-poisoned]

ngimel commented on torch/utils/benchmark/utils/timer.py:
```diff
@@ -172,6 +179,7 @@ def __init__(
         self,
         stmt: str = "pass",
         setup: str = "pass",
+        global_setup: Optional[str] = None,
```
very minor nit - why do you need global_setup as Optional[str] instead of str = ""? That's what it ends up being anyway?

robieta replied:
That's a good point. I'm probably getting overzealous with optionals when here the null state actually is empty string.

ngimel reviewed (approved):
lgtm

Commit:
Update on "Add global setup to timer to allow #include, etc."


Instruction count benchmarks need some includes for IValues, but this is also just generally useful. (Unlike Python where you can just drop imports anywhere, C++ will get very upset if you `#include` in a function body...)

[ghstack-poisoned]
